sdk/src/client/api/block_builder/input_selection/requirement/native_tokens.rs

```diff
@@ -1,131 +1,121 @@
-                }
-            }
-        }
-    }
-
-    if native_tokens_diff.is_empty() {
-        Ok(None)
-    } else {
-        Ok(Some(native_tokens_diff.finish()?))
-    }
-}
-
 impl InputSelection {
     pub(crate) fn fulfill_native_tokens_requirement(&mut self) -> Result<Vec<InputSigningData>, Error> {
-        let mut input_native_tokens = get_native_tokens(self.selected_inputs.iter().map(|input| &input.output))?;
-        let mut output_native_tokens = get_native_tokens(self.non_remainder_outputs())?;
-        let (minted_native_tokens, melted_native_tokens) = self.get_minted_and_melted_native_tokens()?;
-
-        input_native_tokens.merge(minted_native_tokens)?;
-        output_native_tokens.merge(melted_native_tokens)?;
+        let (input_nts, output_nts) = self.get_input_output_native_tokens();
+        let diffs = get_native_tokens_diff(output_nts, input_nts);
+        if diffs.is_empty() {
+            log::debug!("Native tokens requirement already fulfilled");
 
-        if let Some(burn) = self.burn.as_ref() {
-            output_native_tokens.merge(NativeTokensBuilder::from(burn.native_tokens.clone()))?;
+            return Ok(Vec::new());
         }
 
-        // TODO weird that it happens in this direction?
-        if let Some(diffs) = get_native_tokens_diff(&output_native_tokens, &input_native_tokens)? {
-            log::debug!(
-                "Fulfilling native tokens requirement with input {input_native_tokens:?} and output {output_native_tokens:?}"
-            );
-
-            let mut newly_selected_inputs = Vec::new();
-            let mut newly_selected_ids = HashSet::new();
-
-            for diff in diffs.iter() {
-                let mut amount = U256::zero();
-                // TODO sort ?
-                let inputs = self.available_inputs.iter().filter(|input| {
-                    input
-                        .output
-                        .native_token()
-                        .is_some_and(|native_token| native_token.token_id() == diff.token_id())
-                });
-
-                for input in inputs {
-                    amount += input
-                        .output
-                        .native_token()
-                        // PANIC: safe to unwrap as the filter guarantees inputs containing this native token.
-                        .unwrap()
-                        .amount();
-
-                    if newly_selected_ids.insert(*input.output_id()) {
-                        newly_selected_inputs.push(input.clone());
-                    }
-
-                    if amount >= diff.amount() {
-                        break;
-                    }
+        log::debug!("Fulfilling native tokens requirement");
+
+        let mut newly_selected_inputs = Vec::new();
+        let mut newly_selected_ids = HashSet::new();
+
+        for (&token_id, &amount) in diffs.iter() {
+            let mut input_amount = U256::zero();
+            // TODO sort ?
```

Comment:
Will this be addressed in this PR? If not, do we have an issue for this already we can reference here?

Reply:
Not even really sure what the best solution is. I guess we could create a mapping of the tokens to the outputs so we don't have to iterate every time. But yeah, separate issue for sure.